lib/torrent.js

```diff
@@ -1433,62 +1402,65 @@ class Torrent extends EventEmitter {
       const req = minWire.requests[i]
       if (req.piece !== index) continue
 
-      self.pieces[index].cancel((req.offset / Piece.BLOCK_LENGTH) | 0)
+      this.pieces[index].cancel((req.offset / Piece.BLOCK_LENGTH) | 0)
     }
 
-    self.emit('hotswap', minWire, wire, index)
+    this.emit('hotswap', minWire, wire, index)
     return true
   }
 
   /**
    * Attempts to request a block from the given wire.
    */
   _request (wire, index, hotswap) {
-    const self = this
     const numRequests = wire.requests.length
     const isWebSeed = wire.type === 'webSeed'
 
-    if (self.bitfield.get(index)) return false
+    if (this.bitfield.get(index)) return false
 
     const maxOutstandingRequests = isWebSeed
       ? Math.min(
-        getPiecePipelineLength(wire, PIPELINE_MAX_DURATION, self.pieceLength),
-        self.maxWebConns
+        getPiecePipelineLength(wire, PIPELINE_MAX_DURATION, this.pieceLength),
+        this.maxWebConns
       )
       : getBlockPipelineLength(wire, PIPELINE_MAX_DURATION)
 
     if (numRequests >= maxOutstandingRequests) return false
-    // var endGame = (wire.requests.length === 0 && self.store.numMissing < 30)
+    // var endGame = (wire.requests.length === 0 && this.store.numMissing < 30)
 
-    const piece = self.pieces[index]
+    const piece = this.pieces[index]
     let reservation = isWebSeed ? piece.reserveRemaining() : piece.reserve()
 
-    if (reservation === -1 && hotswap && self._hotswap(wire, index)) {
+    if (reservation === -1 && hotswap && this._hotswap(wire, index)) {
       reservation = isWebSeed ? piece.reserveRemaining() : piece.reserve()
     }
     if (reservation === -1) return false
 
-    let r = self._reservations[index]
-    if (!r) r = self._reservations[index] = []
+    let r = this._reservations[index]
+    if (!r) r = this._reservations[index] = []
     let i = r.indexOf(null)
     if (i === -1) i = r.length
     r[i] = wire
 
     const chunkOffset = piece.chunkOffset(reservation)
     const chunkLength = isWebSeed ? piece.chunkLengthRemaining(reservation) : piece.chunkLength(reservation)
 
-    wire.request(index, chunkOffset, chunkLength, function onChunk (err, chunk) {
-      if (self.destroyed) return
+    wire.request(index, chunkOffset, chunkLength, async (err, chunk) => {
+      if (this.destroyed) return
+
+      const onUpdateTick = () => {
+        process.nextTick(() => { this._update() })
+      }
 
       // TODO: what is this for?
-      if (!self.ready) return self.once('ready', () => { onChunk(err, chunk) })
+      if (!this.ready) await new Promise(resolve => this.once('ready', resolve))
```

Comment:
I used await and an async function here to allow this callback to be an anonymous arrow function. I think that's OK because async functions are supported in every browser that supports webRTC, and in node 8 which is LTS. (the tests wouldn't pass otherwise as they're set to the latest LTS in travis)